modules/apps/portal/portal-company-log/src/main/java/com/liferay/portal/company/log/internal/servlet/CompanyLogServlet.java

```diff
@@ -0,0 +1,345 @@
+				_downloadLogFile(
+					permissionChecker, httpServletRequest, httpServletResponse,
+					pathArray);
+			}
+		}
+		catch (FileNotFoundException fileNotFoundException) {
+			if (_log.isWarnEnabled()) {
+				_log.warn(fileNotFoundException);
+			}
+
+			_portal.sendError(
+				HttpServletResponse.SC_NOT_FOUND, fileNotFoundException,
+				httpServletRequest, httpServletResponse);
+		}
+		catch (Exception exception) {
+			if (_log.isWarnEnabled()) {
+				_log.warn(exception);
+			}
+
+			_portal.sendError(
+				exception, httpServletRequest, httpServletResponse);
+		}
+	}
+
+	private void _downloadLogFile(
+			PermissionChecker permissionChecker,
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String[] pathArray)
+		throws Exception {
+
+		long companyId = GetterUtil.getLongStrict(pathArray[0]);
+
+		if (_companyLocalService.fetchCompanyById(companyId) == null) {
+			throw new NoSuchCompanyException(
+				"No Company exists with the primary key " + companyId);
+		}
+
+		if (!permissionChecker.isCompanyAdmin(companyId)) {
+			throw new PrincipalException.MustBeCompanyAdmin(
+				permissionChecker.getUserId());
+		}
+
+		String logFilesDirPath = Log4JUtil.getCompanyLogDirectory(companyId);
+
+		if (logFilesDirPath == null) {
+			return;
+		}
+
+		String fileName = pathArray[1];
+
+		Path path = Paths.get(logFilesDirPath, fileName);
+
+		path = path.normalize();
+
+		if (!path.startsWith(logFilesDirPath)) {
+			throw new PrincipalException("Unauthorized access");
+		}
+
+		File logFile = path.toFile();
+
+		if (!logFile.exists()) {
+			throw new FileNotFoundException(
+				StringBundler.concat(
+					"Unable to find log file ", fileName, " for company ",
+					companyId));
+		}
+
+		String startIndex = ParamUtil.getString(
+			httpServletRequest, "startIndex");
+		String endIndex = ParamUtil.getString(httpServletRequest, "endIndex");
+
+		if (Validator.isNull(startIndex) && Validator.isNull(endIndex)) {
+			ServletResponseUtil.sendFile(
+				httpServletRequest, httpServletResponse, fileName,
+				new FileInputStream(logFile), logFile.length(),
+				_mimeTypes.getContentType(fileName),
+				HttpHeaders.CONTENT_DISPOSITION_ATTACHMENT);
+		}
+		else {
+			if (Validator.isNull(startIndex)) {
+				startIndex = "0";
+			}
+
+			int start = GetterUtil.getIntegerStrict(startIndex);
+
+			int logFileLength = (int)logFile.length();
+
+			if (Validator.isNull(endIndex) && (start >= 0)) {
+				endIndex = String.valueOf(logFileLength);
+			}
+
+			int end = GetterUtil.getIntegerStrict(endIndex);
+
+			if ((start < 0) || (end < 0) || ((start == 0) && (end == 0)) ||
+				(start >= end)) {
+
+				throw new PrincipalException(
+					"startIndex or endIndex can not be less than 0, and " +
+						"startIndex can not be greater than or equal to " +
+							"endIndex");
+			}
+
+			if (end > logFileLength) {
+				end = logFileLength;
+			}
+
+			if (start != 0) {
+				--start;
+			}
+
+			byte[] bytes = new byte[logFileLength];
```

Comment:
@yuhai Reading everything into memory may cause out of memory issue, please use stream here.